packer_test/common/plugin.go

```diff
@@ -77,6 +132,83 @@ func (ts *PackerTestSuite) GetPluginPath(t *testing.T, version string) string {
 	return path.(string)
 }
 
+type CompilationResult struct {
+	Error   error
+	Version string
+}
+
+// Ready processes a series of CompilationResults, as returned by CompilePlugin
+//
+// If any of the jobs requested failed, the test will fail also.
+func Ready(t *testing.T, results []chan CompilationResult) {
+	for _, res := range results {
+		jobErr := <-res
+		empty := CompilationResult{}
+		if jobErr != empty {
+			t.Errorf("failed to compile plugin at version %s: %s", jobErr.Version, jobErr.Error)
+		}
+	}
+
+	if t.Failed() {
+		t.Fatalf("some plugins failed to be compiled, see logs for more info")
+	}
+}
+
+type compilationJob struct {
+	versionString  string
+	suite          *PackerTestSuite
+	done           bool
+	resultCh       chan CompilationResult
+	customisations []BuildCustomisation
+}
+
+// CompilationJobs keeps a queue of compilation jobs for plugins
+//
+// This approach allows us to avoid conflicts between compilation jobs.
+// Typically building the plugin with different ldflags is safe to perform
+// in parallel on the same file set, however customisations tend to be more
+// conflictual, as two concurrent compilation jobs may end-up compiling the
+// wrong plugin, which may cause some tests to misbehave, or even compilation
+// jobs to fail.
+//
+// The solution to this approach is to have a global queue for every plugin
+// compilation to be performed safely.
+var CompilationJobs = struct {
+	Mutex   *sync.Mutex
+	Jobs    chan compilationJob
+	Running bool
+}{
+	Mutex:   &sync.Mutex{},
+	Jobs:    make(chan compilationJob, 10),
+	Running: false,
+}
+
+func processJobs() {
+	CompilationJobs.Mutex.Lock()
+	defer CompilationJobs.Mutex.Unlock()
+
+	if CompilationJobs.Running {
+		return
+	}
+
+	go func() {
+		for job := range CompilationJobs.Jobs {
+			log.Printf("compiling plugin on version %s", job.versionString)
+			err := compilePlugin(job.suite, job.versionString, job.customisations...)
+			if err != nil {
+				job.resultCh <- CompilationResult{
+					Error:   err,
+					Version: job.versionString,
+				}
+			}
+			close(job.resultCh)
+		}
+		CompilationJobs.Mutex.Lock()
+		defer CompilationJobs.Mutex.Unlock()
+		CompilationJobs.Running = false
+	}()
+}
+
```

Comment:
I think for this kind of queue, I like better having a goroutine that would listen to the Jobs channel instead, that way you could get rid of the mutexes

Reply:
As discussed out-of-band, I agree that the original code was a bit messy with the mutex and the running flag; these were there so we could detect when no jobs were queued and we could stop the coroutine in that case.

There were two main problems with the code though:

1. The logic has a small likelihood of deadlocking; if at one point a coroutine was exiting the loop while another was queuing a job, it would've been possible to queue a job, and place the state as non-running, leaving one compilation job that may not be picked-up. This never occurred and the timing is a bit tight to be in this situation, but it was possible nonetheless.
2. The detection of the completion was borked since the jobs channel was never closed, so the coroutine never effectively left the loop for consuming incoming jobs, so all that flawed logic was not even necessary, as effectively the channel and coroutines were persistent, so we can completely do away with the `Running` flag and the mutex.

I've rerolled a change on that code already, which does away with this logic, in favour of having the coroutine started with the process at initialisation, and never cleaning it up.
Since this is all happening in a `_test` package, it won't be linked to executables other than tests, so we should be covered on this front.